org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/hover/JvmAnnotationReferencePrinter.java

```diff
@@ -0,0 +1,220 @@
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.xtext.common.types.JvmAnnotationReference;
+import org.eclipse.xtext.common.types.JvmAnnotationType;
+import org.eclipse.xtext.common.types.JvmAnnotationValue;
+import org.eclipse.xtext.common.types.JvmOperation;
+import org.eclipse.xtext.ui.editor.hover.html.XtextElementLinks;
+import org.eclipse.xtext.xbase.XAbstractFeatureCall;
+import org.eclipse.xtext.xbase.XBinaryOperation;
+import org.eclipse.xtext.xbase.XBooleanLiteral;
+import org.eclipse.xtext.xbase.XFeatureCall;
+import org.eclipse.xtext.xbase.XListLiteral;
+import org.eclipse.xtext.xbase.XMemberFeatureCall;
+import org.eclipse.xtext.xbase.XNumberLiteral;
+import org.eclipse.xtext.xbase.XStringLiteral;
+import org.eclipse.xtext.xbase.XTypeLiteral;
+import org.eclipse.xtext.xbase.annotations.xAnnotations.XAnnotation;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+
+/**
+ * Provides a HTML representation for {@code JvmAnnotationReference}s to be used in the hover.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public class JvmAnnotationReferencePrinter {
+
+	/**
+	 * @return the HTML representation for the passed {@code reference}
+	 * @throws NullPointerException
+	 *             if the {@code reference} is {@code null}
+	 */
+	public String toHtmlString(JvmAnnotationReference reference) {
+		Preconditions.checkNotNull(reference);
+		return internalToString(reference);
+	}
+
+	protected String internalToString(Object obj) {
+		if (obj instanceof XBinaryOperation) {
+			return _internalToString((XBinaryOperation) obj);
+		} else if (obj instanceof XFeatureCall) {
+			return _internalToString((XFeatureCall) obj);
+		} else if (obj instanceof XListLiteral) {
+			return _internalToString((XListLiteral) obj);
+		} else if (obj instanceof XMemberFeatureCall) {
+			return _internalToString((XMemberFeatureCall) obj);
+		} else if (obj instanceof XBooleanLiteral) {
+			return _internalToString((XBooleanLiteral) obj);
+		} else if (obj instanceof XNumberLiteral) {
+			return _internalToString((XNumberLiteral) obj);
+		} else if (obj instanceof XStringLiteral) {
+			return _internalToString((XStringLiteral) obj);
+		} else if (obj instanceof XTypeLiteral) {
+			return _internalToString((XTypeLiteral) obj);
+		} else if (obj instanceof XAnnotation) {
+			return _internalToString((XAnnotation) obj);
+		} else if (obj instanceof JvmAnnotationReference) {
+			return _internalToString((JvmAnnotationReference) obj);
+		} else if (obj instanceof JvmAnnotationValue) {
+			return _internalToString((JvmAnnotationValue) obj);
+		} else {
+			return _internalToString(obj);
+		}
+	}
+
+	protected String _internalToString(JvmAnnotationReference reference) {
+		StringBuilder buffer = new StringBuilder();
+		buffer.append("@");
+		URI uri = EcoreUtil.getURI(reference.getAnnotation());
```

Comment:
I have added a private method `#createLinkToOperation(...)` which abstract a bit of the common logic from the lines pointed above. The method has a minor optimization - the logic to filter out the first operation with simple-name `value` is only executed if the `explicitOperation` is `null`. 

Secondly, I removed the `scheme` parameter from the method `createLinkWithLabel` and hard-coded the scheme to `XtextElementLinks.XTEXTDOC_SCHEME`. This removes a bit of repetitive usage of the constant, secondly, since this is private code which is mostly controlled, I think it is ok to make such a change. @szarnekow What do you think? Is ok to make such a change?